### DISCUSSION TITLE
Multiple block flows.

### DISCUSSION THREAD

**Discussion by TekyTeky** [Novakasa/brickrail#118] [Ideas]
I had an Idea or two... if blocks were labeled then you could have a Block Flow 
eg:  Block 1 was a station named "City", block 2 was a coal yard, block 3 was another station called "Central" maybe on a layer below.
You could have it go from Block 1 to Block 3 then block 2 and then stop.
Even better would be to have delays on a block, eg a station block could have a delay of 10 seconds then continue.
Just some ideas to make it hands free.

**Comment by Novakasa** [comment]
Yeah something like that was also on my mind. I was thinking of a concept of "block groups". E.g. the "city" group could consist of multiple blocks that represent multiple platforms in the city train station.
Then a train schedule could be programmed to reach certain groups at certain times, while the specific block will be chosen dynamically depending on which block is not occupied.

**Reply by TekyTeky** [reply]
Wow, that would be great.. glad you have been thinking about it too. 😎

**Comment by TekyTeky** [comment]
Had you had any thoughts about people having long rakes (consists) of trains? That they would overflow into other blocks?

What if you had a "marker" that indicated a block was occupied?
I ask because it is a problem for model railway people. But as their track is powered, they put a resister in the rolling stock and as it draws current the train app system thinks it is occupied... similarly if you are shunting and have left wagons in a section of track, you could "Mark" it as "In Use" and it would be avoided by the system.
These are things that I would like, I guess.

Also, some people would only be running a small track like the test one I have for Brickrail or are you really aiming at layouts the size of yours, which would be a reasonable assumption?

The other option that you may not want to even consider is that I have a lot of WeDo2 Lego distance sensors (they are better for distance sensing than the color ones, for BAP anyway, as I have said previously.
 It would be good to be able to plug one in as some type of track sensor, eg for initiating rail crossings etc.
 
 I have a Pybricks one that I run now, and it works nicely, although it was difficult getting the timings right.
A train goes past a sensor and initiates the code and both boom gates drop, then after a predetermined time raise. I probably should have another sensor to raise them. Then the train sensor sees a color at some point before on the track and a .wav file plays the bells of a rail crossing.

Anyway, food for thought, just to help you sleep...not...   ;-P

**Reply by Novakasa** [reply]
> What if you had a "marker" that indicated a block was occupied?

I think the "marker" system for detecting block overflows is not going to work, since the train doesn't really know how long it is (and brickrail doesn't know the real size of blocks). The easiest way would probably for the user to set these sizes. However, the way that Blocks in brickrail work (and also in Rocrail I think), it is undefined which parts of the layout should then actually be prohibited for other trains.

The whole concept of blocks for train automation relies on the assumption that a block can contain every train _fully_, so if that assumption is relaxed, the whole navigation system would probably need to be remade. For this reason, I currently don't plan to implement things like that.

> Also, some people would only be running a small track like the test one I have for Brickrail or are you really aiming at layouts the size of yours, which would be a reasonable assumption?

My goal, I think, is to have automation that is _simple_ but still allows for complex scale. I plan to stick to the blocks system, because it is simple, but since the trains have their own sensors, the amount of blocks is theoretically unlimited. This means that you can have quite complex layouts (like in my video) assuming that you can find a way to motorize all the required switches and don't run into bluetooth issues with too many hubs.

This is why I've opted to for now focus on sensors on the trains, rather than stationary sensors, since then it is much easier to have many blocks. You would probably need at least two stationary sensors per block to have similar level of automation as the current system. Of course, if you plan to have a collection of 20 trains, then it becomes difficult to equip all of them with a color sensor :sweat_smile: .

I never planned for a specific "scale" for automation, I started quite small and since then made the layouts more complex, and basically still am waiting for when I will run into bluetooth issues. My complex layouts managed to have 5 hubs at the same time, which is the most hubs I have ever tried, and I was quite surprised that it seemed to work without any big issues.

> A train goes past a sensor and initiates the code and both boom gates drop, then after a predetermined time raise.

I have thought about things like that as well. These probably don't need additional sensors. You could have regular markers on the tracks which then could trigger an "action" like playing a sound or closing rail crossing booms.

Rail crossings could also be implemented by assigning them to a location in brickrail, and if ever a train "reserves" this part of the layout, the crossing booms would automatically go down, and go up again once the train has "released" the lock on this track.

Crossings would then also be connected to the layout controllers. One could also imagine attaching signals to layout controllers, however there is currently no good way to have complex (RGB?) lights with PoweredUp. Motorized signals could work though, but that would be quite big with PoweredUp motors.

For now, crossings would be the first additional device to try to implement.

**Reply by Novakasa** [reply]
BTW, you might have noticed in my "layers" layout I don't have every switch motorized in the upper layer. I am using "one way" tracks to make every train guarantee to only go certain routes, such that some switches never need to be actually switched, if they stay in one predetermined position. At some point I should write about this possibility on the Wiki.

**Reply by Novakasa** [reply]
Hmm, maybe overflow is not impossible, if we compare train size to block size (set by the user) and the train doesn't fit, we could simply lock every other block connected in the direction the train came from. This would make the most sense if the user sets up "overflow" blocks, so that the train then only locks two blocks at most.

This would still be a lot of work, since there is no concept yet of locking an arbitrary number of blocks and their connecting track.

**Reply by TekyTeky** [reply]
Yeah I thought that would be what you thought.. I just put it out their about long trains.
My future layout would be more like yours, cleaner.

If plate sensors could be used as detectors that would be great.. really great.
My train signals are leds from a local Lego Lights supplier, I make the signals myself and have a magnet on the switch and a reed switch.
Once I build then into the layout you don't see any wires and the switches have huts build on them etc.

<img width="723" alt="image" src="https://user-images.githubusercontent.com/87248517/236045018-a1d38938-d2e9-4372-b188-13e472769cb1.png">

Im my testing, and reading a lot on Eurobricks, most basic Windows PCs have a BT capability of 7 BLE devices.
The more expensive the machine (relating to better BLE components I guess) the more you get.
My basic old Windows pc get's 7, My Mac running Bootcamp (a windows emulator) gets 8.
My current Lenovo 8Gen i7 16gb of ram I can allocate 10 BLE devices.. hence I use the Technic hubs for sensors and Switches... the rest are trains.. I can only run 3 trains at any one time as the rest are layout hubs.

I tried BLE USB devices but couldn't get them to work, but the guy who wrote BAP used a Mac with Bootcamp and had one and got 15 devices to connect... but I tried with the same one and couldn't get Windows to recognise it instead of the build in one.

I did notice your layers, and the flow, hence my previous comment about Trixbrix... I may just have to bite the bullet and make an order from the EU when I rebuild my layout.
There is a guy on YT and Insta called SevBrick, he has a huge layout and all the trains interweave like yours but on a larger scale. 
https://instagram.com/sevbrick?igshid=YmMyMTA2M2Y=

**Reply by Novakasa** [reply]
The signals look really cool!

Though for now I am probably limiting features to mainly support official LEGO hardware, as it has been the Brickrail philosophy from the beginning, makes it much easier for people to set up their layout (Also I haven't really made the jump to experiment with custom electronics yet).

Have you built any motorized rail crossings? How should the motor be operated? Should we rely on motors with rotation sensors to have the booms "angle calibrated" or should they rely on a clutch gear or similar with just a constant "time" of turning the motor?

**Reply by TekyTeky** [reply]
Maybe you didn't see a previous post where I said I made one from Pybricks sw.
The trains go past a sensor (which in Brickrail's case could be a colored plate) Mine is a Lego distance sensor which are still available for education and sometime Lego put them into kits, like the grand piano.

It has a Technic hub powering it and 2 motors, but with gears you could use one. I investigated so many ideas from users but as I had the motors used 2, it was also easier to position them.

The motors are timed, similar to your switches, not rotation motors, I use the older WeDo2 motors that can't be set as rotation. They have a little brick at the end that looks realistic. If you look into new PU motors, they are all able to do rotations themselves... but it was hard to find out how... PyBricks have an article now on rotation motors.

Anyway, a train goes past the sensor and the booms go down for a predetermined time... and Lego lights flash too above the gate end... also powered by the same hub and Pybricks. In the old Pybricks I couldn't get it to stay down longer or slower. And had no one to help me.
As all ports were full on the hub, I had no way to add a raise boom sensor.

**Reply by TekyTeky** [reply]
Here is the code I used.. bear in mind it is 2 years old.

[RailCrossingPY2.zip](https://github.com/Novakasa/brickrail/files/11402160/RailCrossingPY2.zip)

**Comment by TekyTeky** [comment]
It's crude when you see it here, so maybe you can write some quick code. ;-)
https://user-images.githubusercontent.com/87248517/236331338-87ea24d0-294c-42e3-9037-552e77d05bf2.mov

**Reply by TekyTeky** [reply]
Not sure if you saw the above re rail crossings.

**Reply by Novakasa** [reply]
Saw it :) Rail crossings are definitely planned, but I'll probably wait until I'm at my parents place where I have the layout setup
